src/components/tooltip/Tooltip.tsx

```diff
@@ -0,0 +1,306 @@
+import omit from 'lodash.omit';
+import React, { MouseEventHandler, ReactNode, useEffect, useMemo, useRef, useState } from 'react';
+import { createPortal } from 'react-dom';
+import Transition from 'react-transition-group/Transition';
+import { GenericComponent } from '../../@types/types';
+import { COLORS, SIZES } from '../../constants';
+import Box from '../box';
+import { BoxProps } from '../box/Box';
+import DocumentObjectProvider, { Context as DocumentObjectContext } from '../hoc/DocumentObjectProvider';
+import { getViewport } from '../utils/utils';
+import theme from './theme.css';
+
+type Position = 'bottom' | 'horizontal' | 'left' | 'right' | 'top' | 'vertical';
+
+export const POSITIONS: Record<string, Position> = {
+  BOTTOM: 'bottom',
+  HORIZONTAL: 'horizontal',
+  LEFT: 'left',
+  RIGHT: 'right',
+  TOP: 'top',
+  VERTICAL: 'vertical',
+};
+
+interface PositionState {
+  position: Position;
+  top: number | string;
+  left: number | string;
+}
+export type AllowedColor = Exclude<typeof COLORS[number], 'teal'> | 'white' | 'inverse';
+export type AllowedSize = Exclude<typeof SIZES[number], 'tiny' | 'fullscreen' | 'smallest' | 'hero'>;
+const SIZE_MAP: Record<AllowedSize, BoxProps> = {
+  large: {
+    padding: 4,
+  },
+  medium: {
+    padding: 3,
+  },
+  small: {
+    paddingHorizontal: 3,
+    paddingVertical: 2,
+  },
+};
+
+interface TooltippedComponentProps {
+  children: ReactNode;
+  className?: string;
+  onClick?: MouseEventHandler;
+  onMouseEnter?: MouseEventHandler;
+  onMouseLeave?: MouseEventHandler;
+  onTooltipEntered?: () => void;
+  tooltip: ReactNode;
+  tooltipColor?: AllowedColor;
+  tooltipHideOnClick?: boolean;
+  tooltipIcon?: ReactNode;
+  tooltipPosition?: Position;
+  tooltipShowOnClick?: boolean;
+  tooltipSize?: AllowedSize;
+  documentObject: Document;
+  tooltipShowDelay?: number;
+  /** The z-index of the Tooltip */
+  zIndex?: number;
+  tooltipActive?: boolean;
+  ComposedComponent: React.ElementType;
+  // Because its HOC, it needs to pass props of child component
+  [key: string]: any;
+}
+
+export interface TooltipProps extends Omit<TooltippedComponentProps, 'ComposedComponent'> {}
+
+const TooltippedComponent: GenericComponent<TooltippedComponentProps> = ({
+  children,
+  className,
+  documentObject,
+  tooltip,
+  tooltipColor = 'white',
+  onTooltipEntered,
+  tooltipHideOnClick = true,
+  tooltipIcon = null,
+  tooltipPosition = 'top',
+  tooltipShowOnClick = false,
+  tooltipSize = 'medium',
+  tooltipShowDelay = 100,
+  zIndex = 700,
+  onClick,
+  onMouseEnter,
+  onMouseLeave,
+  tooltipActive = null,
+  ComposedComponent,
+  ...other
+}) => {
+  const tooltipRoot = useMemo(() => documentObject.createElement('div'), []);
+  const ref = useRef(null);
+  const [active, setActive] = useState(false);
+  const [position, setPosition] = useState<PositionState>({ position: tooltipPosition, top: 'auto', left: 'auto' });
+
+  const activate = (position: PositionState | undefined) => {
```

Comment:
It's a bit hard to review since so much has changed but you had to add all the handling for null/undefined right? Would it be an option to use casts (or `!`) to remove null/undefined since apparently it never happened in the original code?

Reply:
yes, it wasn't correctly re-implemented